### PR TITLE
Modify user serializer to add gender

### PR DESCRIPTION
**Pull request by JorgeMichelena**
-Modified User model to add 'gender' field
-Changed default serializer for user model to custom serializer with 'gender' field
-Created test cases to test edit profile functionality
-Created UserFactory class to use it in test cases

### CODE REVIEW

File: api/tests.py [hunk 1]
```diff
@@ -1,3 +1,100 @@
...
 
-# Create your tests here.
+class EditProfileTests(TestCase):
+        
+    def setUp(self):
+        self.request_factory = APIRequestFactory()
+        self.user1 = UserFactory()
+        self.user2 = UserFactory()
+        self.user1.save()
+        self.user2.save()
```
Comment (kaozdl) [inline api/tests.py]: why are you saving the users?

File: api/serializers.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+from rest_framework import serializers
+from users.models import User
+
+class UserDetailsSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = User
+        fields = ('pk', 'username', 'email', 'first_name', 'last_name', 'gender')
```
Comment (kaozdl) [inline api/serializers.py]: if more than 3 fields try to add one field per line.

File: api/tests.py [hunk 3]
```diff
@@ -1,3 +1,100 @@
...
+        self.view = UserDetailsView.as_view()
+    
+    def test_logged_user_changes_not_read_only_fields(self):
```
Comment (GuilleLeopold) [inline api/tests.py]: editable_fields would be fine I guess

File: users/factory.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+class UserFactory(factory.Factory):
+    class Meta:
+        model = User
+    
+    username = factory.Sequence(lambda n: 'user%d' % n)
+    email = factory.LazyAttribute(lambda obj: '%s@example.com' % obj.username)
+    first_name = faker.Faker().first_name() 
+    last_name = faker.Faker().last_name()
+    password = '1234'
```
Comment (GuilleLeopold) [inline users/factory.py]: is there any way to fake the password?

File: api/serializers.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+
+from rest_framework import serializers
+from users.models import User
+
+class UserDetailsSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = User
+        fields = ('pk',
```
Comment (jiss2891) [inline api/serializers.py]: I don't know if there is a preferred way in the company, I do prefer this kind of indentation:

```python
fields = (
    'pk', 
    'username', 
    'email', 
    'first_name', 
    'last_name', 
    'gender',
)
```

I show this to you just as an alternative. Notice the last comma, after 'gender'; if you do so, when adding new elements, the git's diff will only show the lines where the items were added. It improves readability in code reviews.

File: api/serializers.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (jiss2891) [inline api/serializers.py]: This blank line is not necessary.

File: api/tests.py [hunk 7]
```diff
@@ -1,3 +1,103 @@
+from users.factory import UserFactory
 from django.test import TestCase
+from api import serializers
+from users.models import User
+from rest_framework.test import APIClient, force_authenticate, APIRequestFactory
+from rest_auth.views import UserDetailsView
 
-# Create your tests here.
+class EditProfileTests(TestCase):
+
```
Comment (jiss2891) [inline api/tests.py]: Neither this one.

File: api/tests.py [hunk 8]
```diff
@@ -1,3 +1,103 @@
...
+    def test_logged_user_changes_editable_fields(self):
+        # arrange
+        old_first_name = self.user1.first_name
+        old_last_name = self.user1.last_name
+        old_gender = self.user1.gender        
+        data = {'username':self.user1.username, 
+                'first_name':'new_first_name', 
+                'last_name':'new_last_name', 
+                'gender':'F'}
+        request = self.request_factory.put('/api/v1', data, format='json')
```
Comment (jiss2891) [inline api/tests.py]: be careful with the missing slash at the end of the URL.

File: api/tests.py [hunk 9]
```diff
@@ -1,3 +1,103 @@
...
+        force_authenticate(request, self.user1)
+        # act
+        response = self.view(request)
+        # assert
+        new_first_name = self.user1.first_name
```
Comment (jiss2891) [inline api/tests.py]: This three assigns are not needed, you can use the self.{thing} inside the assertEquals

File: api/tests.py [hunk 10]
```diff
@@ -1,3 +1,103 @@
...
+        new_last_name = self.user1.last_name
+        new_gender = self.user1.gender
+        
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(new_first_name, 'new_first_name')
+        self.assertEqual(new_last_name, 'new_last_name')
+        self.assertEqual(new_gender, 'F')
+        self.assertFalse(new_first_name==old_first_name)
```
Comment (jiss2891) [inline api/tests.py]: Those assertFalse are not needed, because nothing has changed since the assertEquals.

File: api/tests.py [hunk 11]
```diff
@@ -1,3 +1,103 @@
...
+        self.assertFalse(new_last_name==old_last_name)
+        self.assertFalse(new_gender==old_gender)
+        
+    
+    def test_logged_user_changes_pk(self):
+        # arrange
```
Comment (jiss2891) [inline api/tests.py]: Avoid comments that explain less than the code itself, instead try to say why are you doing this.

File: api/tests.py [hunk 12]
```diff
@@ -1,3 +1,103 @@
...
+        self.request_factory = APIRequestFactory()
+        self.user1 = UserFactory()
+        self.view = UserDetailsView.as_view()
+    
+    def test_logged_user_changes_editable_fields(self):
+        # arrange
+        old_first_name = self.user1.first_name
+        old_last_name = self.user1.last_name
+        old_gender = self.user1.gender        
+        data = {'username':self.user1.username,
```
Comment (jiss2891) [inline api/tests.py]: You can adopt the same indentation approach here

File: users/factory.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import factory
+from factory.faker import faker
+from users.models import User
+
+
+class UserFactory(factory.Factory):
+    class Meta:
+        model = User
+    
+    pk = factory.Sequence(lambda n: n)
```
Comment (jiss2891) [inline users/factory.py]: Class variables have to be at the top of the class definition

File: api/tests.py [hunk 14]
```diff
@@ -1,3 +1,103 @@
...
+        old_pk = self.user1.pk
+        data = {'pk':'1000', 'username':self.user1.username}
+        request = self.request_factory.put('/api/v1', data, format='json')
+        force_authenticate(request, self.user1)
+        # act
+        response = self.view(request)
+        # assert
+        new_pk = self.user1.pk
+        
+        self.assertEqual(response.status_code, 200)
```
Comment (brunomichetti) [inline api/tests.py]: When you check status codes, you can use this https://www.django-rest-framework.org/api-guide/status-codes/ .
It's just a definition of constants. This improves code quality making the code more readable. For example, you can change 200 for HTTP_200_OK, and so on.

**Pull request merged**